samples/extensions/dynamic_primitive_clipping/dynamic_primitive_clipping.cpp

```diff
@@ -0,0 +1,384 @@
+	}
+}
+
+void DynamicPrimitiveClipping::build_command_buffers()
+{
+	VkCommandBufferBeginInfo command_buffer_begin_info = vkb::initializers::command_buffer_begin_info();
+
+	// Clear color and depth values.
+	VkClearValue clear_values[2];
+	clear_values[0].color        = {{0.1f, 0.1f, 0.1f, 0.0f}};
+	clear_values[1].depthStencil = {1.0f, 0};
+
+	// Begin the render pass.
+	VkRenderPassBeginInfo render_pass_begin_info    = vkb::initializers::render_pass_begin_info();
+	render_pass_begin_info.renderPass               = render_pass;
+	render_pass_begin_info.renderArea.offset.x      = 0;
+	render_pass_begin_info.renderArea.offset.y      = 0;
+	render_pass_begin_info.renderArea.extent.width  = width;
+	render_pass_begin_info.renderArea.extent.height = height;
+	render_pass_begin_info.clearValueCount          = 2;
+	render_pass_begin_info.pClearValues             = clear_values;
+
+	for (int32_t i = 0; i < draw_cmd_buffers.size(); ++i)
+	{
+		auto cmd = draw_cmd_buffers[i];
+
+		// Begin command buffer.
+		vkBeginCommandBuffer(cmd, &command_buffer_begin_info);
+
+		// Set framebuffer for this command buffer.
+		render_pass_begin_info.framebuffer = framebuffers[i];
+		// We will add draw commands in the same command buffer.
+		vkCmdBeginRenderPass(cmd, &render_pass_begin_info, VK_SUBPASS_CONTENTS_INLINE);
+
+		// Bind the graphics pipeline.
+		vkCmdBindPipeline(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, sample_pipeline);
+
+		// Set viewport dynamically
+		VkViewport viewport = vkb::initializers::viewport(static_cast<float>(width), static_cast<float>(height), 0.0f, 1.0f);
+		vkCmdSetViewport(cmd, 0, 1, &viewport);
+
+		// Set scissor dynamically
+		VkRect2D scissor = vkb::initializers::rect2D(width, height, 0, 0);
+		vkCmdSetScissor(cmd, 0, 1, &scissor);
+
+		// Enable primitive clipping dynamically (if required)
+		if (params.useClipping)
+			vkCmdSetDepthClipEnableEXT(cmd, VK_TRUE);
+		else
+			vkCmdSetDepthClipEnableEXT(cmd, VK_FALSE);
+
+		// Draw object once using descriptor_positive
+		if (params.drawObject[0])
+		{
+			vkCmdBindDescriptorSets(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_layouts.models, 0, 1, &descriptor_sets.descriptor_positive, 0, NULL);
+			draw_model(models.objects[models.object_index], cmd);
+		}
+
+		// Draw the same object for the second time, but this time using descriptor_negative
+		// Skip second rendering if clipping is disabled.
+		if (params.drawObject[1] && params.useClipping)
+		{
+			vkCmdBindDescriptorSets(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_layouts.models, 0, 1, &descriptor_sets.descriptor_negative, 0, NULL);
+			draw_model(models.objects[models.object_index], cmd);
+		}
+
+		// Turn the primitive clipping off, so that it does not interfere with further rendering ( GUI rendering comes next )
+		// It may interfere, because using dynamic states turns Vulkan pipelines into state machines similar to OpenGL.
+		vkCmdSetDepthClipEnableEXT(cmd, VK_FALSE);
```

Comment:
This line shouldn't be necessary as the "draw_ui" has it's own pipeline which sets this state statically from the pipeline.

Reply:
Line removed.